python/pyarrow/io.pxi

```diff
@@ -1402,8 +1421,14 @@ def allocate_buffer(int64_t size, MemoryPool memory_pool=None,
 
 cdef class BufferOutputStream(NativeFile):
     """
+    An output stream which writes data in-memory producing a pyarrow.Buffer
```

Comment:
C++ uses "An output stream that writes to a resizable buffer". That's maybe a bit more generic? But in any case I would leave out the "as a result when get_value is called", as that is going into details for the summary line of the docstring. It could be mentioned as a second paragraph, though.